src/PBase.Test/Collections/PrimitiveObjectPoolTests.cs

```diff
@@ -0,0 +1,452 @@
+
+            Assert.NotNull(test3);
+            Assert.Equal(4, resizePool.Size);
+            Assert.Equal(3, resizePool.InUse);
+            Assert.Equal(1, resizePool.PoolCount);
+
+            test4 = resizePool.Obtain();
+
+            Assert.NotNull(test4);
+            Assert.Equal(4, resizePool.Size);
+            Assert.Equal(4, resizePool.InUse);
+            Assert.Equal(0, resizePool.PoolCount);
+
+            Assert.False(resizePool.TryObtain(out test5));
+
+            Assert.Null(test5);
+            Assert.Equal(4, resizePool.Size);
+            Assert.Equal(4, resizePool.InUse);
+            Assert.Equal(0, resizePool.PoolCount);
+
+            test5 = resizePool.Obtain();
+
+            Assert.NotNull(test5);
+            Assert.Equal(5, resizePool.Size);
+            Assert.Equal(5, resizePool.InUse);
+            Assert.Equal(0, resizePool.PoolCount);
+
+            Assert.True(resizePool.TryRelease(test5));
+
+            Assert.Equal(5, resizePool.Size);
+            Assert.Equal(4, resizePool.InUse);
+            Assert.Equal(1, resizePool.PoolCount);
+
+            resizePool.Release(test5);
+
+            Assert.Equal(5, resizePool.Size);
+            Assert.Equal(3, resizePool.InUse);
+            Assert.Equal(2, resizePool.PoolCount);
+
+            resizePool.Release(test5);
+
+            Assert.Equal(5, resizePool.Size);
+            Assert.Equal(2, resizePool.InUse);
+            Assert.Equal(3, resizePool.PoolCount);
+
+            resizePool.Release(test5);
+
+            Assert.Equal(5, resizePool.Size);
+            Assert.Equal(1, resizePool.InUse);
+            Assert.Equal(4, resizePool.PoolCount);
+
+            resizePool.Release(test5);
+
+            Assert.Equal(5, resizePool.Size);
+            Assert.Equal(0, resizePool.InUse);
+            Assert.Equal(5, resizePool.PoolCount);
+
+            resizePool.Release(test5);
+
+            Assert.Equal(6, resizePool.Size);
+            Assert.Equal(0, resizePool.InUse);
+            Assert.Equal(6, resizePool.PoolCount);
+        }
+
+#pragma warning disable CS4014 // Because this call is not awaited, execution of the current method continues before the call is completed
+
+        [Fact]
+        async void TestBlockingPrimitiveObjectPoolObtainRelease()
+        {
+            BlockingPrimitiveObjectPool<TestObjectPoolItem> blockPool = new BlockingPrimitiveObjectPool<TestObjectPoolItem>(3);
+
+            Assert.Equal(3, blockPool.Size);
+            Assert.Equal(0, blockPool.InUse);
+            Assert.Equal(3, blockPool.PoolCount);
+
+            TestObjectPoolItem test = new TestObjectPoolItem();
+
+            Assert.False(blockPool.TryRelease(test));
+
+            Assert.Equal(3, blockPool.Size);
+            Assert.Equal(0, blockPool.InUse);
+            Assert.Equal(3, blockPool.PoolCount);
+
+            Task.Run(() =>
+            {
+                blockPool.Release(test);
+            });
+
+            await Task.Delay(500);
```

Comment:
Probably better to use [Barriers](https://docs.microsoft.com/en-us/dotnet/api/system.threading.barrier?view=netcore-3.1) instead of delays for tests like this. Very easy to start getting failed tests if the delay is not long enough or taking far longer to run tests than is necessary.

Reply:
Switched to using auto reset event